✨ Add support for Trio via AnyIO

graingert opened the pull request:
Fixes #310 

see also https://github.com/encode/starlette/pull/1157

graingert committed:
configure strict pytest options

graingert committed:
open/close files with with

graingert committed:
support trio via anyio with starlette 0.15.0

ArcLightSlavik commented on fastapi/concurrency.py:
```diff
@@ -2,9 +2,6 @@
 
 from starlette.concurrency import iterate_in_threadpool as iterate_in_threadpool  # noqa
 from starlette.concurrency import run_in_threadpool as run_in_threadpool  # noqa
-from starlette.concurrency import (  # noqa
```
Would need to flagged as a breaking change in the change log? (message for tiangolo he makes the change logs)

graingert replied:
starlette walked back on the removal of this function

Kludex commented on pyproject.toml:
```diff
@@ -62,7 +62,8 @@ test = [
     "aiofiles >=0.5.0,<0.6.0",
-    "flask >=1.1.2,<2.0.0"
+    "flask >=1.1.2,<2.0.0",
+    "anyio >=3.1.0,<4",
```
It's an `starlette` requirement, so I guess we don't need it here?

graingert replied:
this depends on `anyio[trio]` now, which isn't a starlette requirement

graingert committed:
WebSocketDisconnect only happens in websocket_connect context manager

* `TestClient.websocket_connect()` now must be used as a context manager:

https://github.com/encode/starlette/blame/119c427474a45bd1297f5adcde864b2874dd8c4f/docs/release-notes.md#L13

graingert committed:
drop pytest-asyncio dep

graingert committed:
Merge branch 'configure-strict-pytest' of github.com:graingert/fastapi into anyio

graingert commented:
currently blocked on https://github.com/encode/starlette/pull/1210

graingert committed:
Merge branch 'master' into anyio

graingert commented:
@tiangolo can you enable my CI builds?

graingert commented on pyproject.toml:
```diff
@@ -32,7 +32,7 @@ classifiers = [
 ]
 requires = [
-    "starlette ==0.14.2",
+    "starlette ==0.15.0",
```
```suggestion
    "starlette ==0.16.0",
```

graingert committed:
Update pyproject.toml

graingert commented on pyproject.toml:
```diff
@@ -62,7 +61,8 @@ test = [
     "aiofiles >=0.5.0,<0.6.0",
-    "flask >=1.1.2,<2.0.0"
+    "flask >=1.1.2,<2.0.0",
+    "anyio[trio] >=3.1.0,<4",
```
```suggestion
    "anyio[trio] >=3.2.1,<4",
```

graingert committed:
bump anyio[trio]

graingert committed:
Merge branch 'master' into anyio

graingert committed:
increase coverage with pytest.fail() # pragma: no cover

graingert committed:
drop async_generator and async_exit_stack deps

graingert committed:
drop aiofiles

graingert committed:
Merge branch 'master' into anyio

graingert committed:
Merge branch 'master' of git://github.com/tiangolo/fastapi into anyio

espetro commented:
Hi! Thanks for such an amazing work 🙌🏼 , I cannot wait to try `trio` with FasAPI. How are we supposed to enable it?

graingert commented:
> Hi! Thanks for such an amazing work 🙌🏼 , I cannot wait to try `trio` with FastAPI. How are we supposed to enable it?

You'll need to pip install this branch from git

ArcLightSlavik commented:
> > Hi! Thanks for such an amazing work 🙌🏼 , I cannot wait to try `trio` with FastAPI. How are we supposed to enable it?
> 
> You'll need to pip install this branch from git

Antony to the rescue 😉 https://www.youtube.com/watch?v=AQrskWh-F5E

dhirschfeld commented:
> You'll need to pip install this branch from git

As a `conda` user it would be great to have this in a release (`conda-forge` only package pypi releases).
@graingert has been keeping this up-to-date with `master` - is there anything stopping it being merged?

adriangb commented:
> is there anything stopping it being merged?

I think the only thing missing is review from @tiangolo

graingert committed:
Merge branch 'master' of git://github.com/tiangolo/fastapi into anyio

tiangolo committed:
✏️ Fix typo

tiangolo committed:
⏪ Revert import/reexport style in concurrency to simplify maintainability

tiangolo committed:
♻️ Simplify and inline re-implementation of asyncio.gather()

tiangolo committed:
📌 Pin Starlette to the gradual next version, minimum that supports AnyIO

to simplify migrations for people. I'll make a release with this, and then a new release with Starlette 0.16.0

tiangolo committed:
🔀 Merge master

tiangolo committed:
🔇 Add pytest filterwarning needed by AnyIO in Python 3.9

tiangolo committed:
🔀 Merge master

tiangolo committed:
➕ Add contextlib2 as a direct dependency during the migration

later it will be required by Starlette 0.16.0

graingert commented:
> I also pinned Starlette to the minimum version that supports AnyIO (`0.15.0`) to make it easier for others to gradually migrate.

You should definitely pin to 0.16.0, the breaking change is around the TestClient(async_backend) kwarg which didn't exist before 0.15.0
